### DISCUSSION TITLE
Package the CLI with a VSCODE extension

### DISCUSSION THREAD

**Discussion by lazkany** [eclipse-langium/langium#1243] [Q&A] [answered]
Hi,

is there a way to package the CLI app with the generated vscode extension? 
I do not want to create an independent package for the CLI app alone, but 
rather a single package that include the generated langium vscode
 extension for my language.

I understand that this question might have been asked in different ways, but 
the answers were not satisfactory maybe because the questions were not direct.
 So, I would like to put this question forward with a clear title.

Langium is doing a great job in all aspect, but when it comes to packaging
 and final steps, tutorials are not extensive as in previous steps.

Many Thanks for the efforts.

**Comment by msujew** [comment, marked as answer]
Hey @lazkany,

it really depends on what you want to do with the CLI/generator/interpreter:
1. Do you want it to run on dirty states of files?
2. Do you want to have the option to also publish it to npm as a standalone tool?
3. Do you want to have it run explicitly or implicitly on document change?

Generally, it's no problem to deploy some additional tooling with the vscode extension. One simple example would be to simply run a generator when the `DocumentBuilder.onBuildPhase(DocumentState.Validated, callback)` callback triggers. You would then implicitly run the generator on every document that has successfully validated.

Another approach would be to register a command in vscode that does exactly that - it would just send a custom notification to the language server to perform the generation, using the dirty state of all documents to generate some code. Both approaches suffer from some performance issues, as the generation will block the main thread. Something that can be circumvented using worker threads.

If you don't need to have it run on dirty states of files, you can just also have the CLI packaged in the vscode extension itself and call `fork` on it in the vscode extension. That basically achieves the same thing as the previous approach (but with better performance), without using the dirty state.

There are many more ways of achieving something like that, as you can probably imagine. These are just things that generally work from the top of my head, but it really depends on your use case.

**Reply by lazkany** [reply]
Thanks a lot that really helped
